components/script/dom/filereadersync.rs

```diff
@@ -87,23 +87,16 @@ impl FileReaderSyncMethods for FileReaderSync {
         Ok(output)
     }
 
-    #[allow(unsafe_code)]
     // https://w3c.github.io/FileAPI/#readAsArrayBufferSyncSection
-    fn ReadAsArrayBuffer(&self, cx: JSContext, blob: &Blob) -> Fallible<NonNull<JSObject>> {
+    fn ReadAsArrayBuffer(&self, cx: JSContext, blob: &Blob) -> Fallible<ArrayBuffer> {
         // step 1
         let blob_contents = FileReaderSync::get_blob_bytes(blob)?;
 
         // step 2
-        unsafe {
-            rooted!(in(*cx) let mut array_buffer = ptr::null_mut::<JSObject>());
-            assert!(ArrayBuffer::create(
-                *cx,
-                CreateWith::Slice(&blob_contents),
-                array_buffer.handle_mut()
-            )
-            .is_ok());
-
-            Ok(NonNull::new_unchecked(array_buffer.get()))
-        }
+
+        rooted!(in(*cx) let mut array_buffer = ptr::null_mut::<JSObject>());
+
+        create_typed_array::<ArrayBufferU8>(cx, &blob_contents, array_buffer.handle_mut())
+            .map_err(|_| Error::JSFailed)
```

Comment:
here we should expect, see https://github.com/servo/servo/pull/31202/files#diff-b439f6e64feec015eba6ad780b7dbd9fd0ea48ba25e6ff87da372099f55b8b0cL99

Reply:
Is it perhaps a bit better to fail gracefully when we can?

Reply:
If we can, yes. 

As per the spec, we return the error from reading the bytes(see https://github.com/servo/servo/pull/31202/files#diff-b439f6e64feec015eba6ad780b7dbd9fd0ea48ba25e6ff87da372099f55b8b0cR93). 

At this point, if something went wrong with the unsafe JS related code, I think it's better to panic. 

Do you have another option in mind?

Reply:
I guess the question is whether or not returning `Error::JSFailed` in that case makes sense. Is there some case where `create_typed_array` can fail and the script crate is still in a recoverable state?

Reply:
So just to be clear, here we are talking about the `Error` returned by `ReadAsArrayBuffer`, and as stated above I don't think returning `Error::JSFailed` makes sense(the only error to return is that potentially returned earlier in this method from `FileReaderSync::get_blob_bytes`).

With regards to `create_typed_array`, which returns an `Result<TypedArray<T, *mut JSObject>, ()>`, I'm not sure about all cases--we are currently going through them one by one--so for now the result seems ok, perhaps later we can remove it and assert inisde the function. 

This answer is a bit long because I wasn't sure whether you meant the one or the other.

Reply:
I think if `create_typed_array` can fail, we should handle that failure gracefully here, especially since this function does account for cases where a failure is possible. A future clean up would be to ensure that `create_typed_array` cannot fail, as you say, in which case we would be able to remove the error handling from this function.

If a recoverable error happens we should try to recover. If the error indicates that the script thread is corrupted or in an unrecoverable state, then it makes sense to panic. In general, Servo panics too often after recoverable errors and often gets stuck, as can be seen by the amount of crashes and timeouts in the WPT suites.

Reply:
Yes I agree we should recover from recoverable cases, in this case I think it is not recoverable, because:

1. We are in the `ArrayBuffer` branch of https://w3c.github.io/FileAPI/#blob-package-data
2. The "Create an ArrayBuffer" operation does not allow for failure, as per https://webidl.spec.whatwg.org/#js-buffer-source-types

If we want to remove ambiguity, we should remove the result from the return of `create_typed_array`. It was an error to include it there(no pun intended).

Reply:
Sorry the link to creating a new buffer is https://webidl.spec.whatwg.org/#arraybuffer-create

By the way the '?' Operator they use made me wonder if it did allow for failure, but on a first reading it doesn't appear to.